### PR TITLE
Fix MeshResourceMarker for mesh with color-based embedded material

### LINKED ISSUE #927: color-based meshes rendering issue
Hello,

I am switching from ROS1 to ROS2 Humble and I noticed that the issue of color-based meshes is reproduced with rviz2.

Impact: Color based meshes are white

On ROS1 RViz issue has been fixed by https://github.com/ros-visualization/rviz/pull/1424

The following repo https://github.com/xbroquer/rviz_collada_marker/tree/ros2  can be used to reproduce the issue.

ros2 launch rviz_collada_marker mesh_marker_test.launch.py

PR2 Base is well rendered but not the moonwalk and nil link

Below the actual result using RViz from Humble:
Screenshot from 2022-11-24 22-47-32

### PR DESCRIPTION
**Pull request by xbroquer**
Issue: Color from embedded color based mesh are overrided and not rendered as expected

It was an known issue on ROS1 that was fixed by
https://github.com/ros-visualization/rviz/pull/1424

Kudos to the original author

Fix #927

Signed-off-by: Xavier BROQUERE <xav.broquere@gmail.com>

### CODE REVIEW

**Comment by xbroquere** [comment]
In order to test this PR , I used this repo https://github.com/xbroquer/rviz_collada_marker/tree/ros2
ros2 launch rviz_collada_marker mesh_marker_test.launch.py

| Humble            |  PR tested on Humble |
:-------------------------:|:-------------------------:
Screenshot from 2022-11-25 09-07-49 | Screenshot from 2022-11-25 09-04-02

**Comment by xbroquere** [comment]
Hello @ahcorde, 

 is it possible to review this PR ?  it will be great if this could be in the coming  humble sync :)

Thanks!

**Review by clalancette — APPROVED** [review]

File: rviz_default_plugins/src/rviz_default_plugins/displays/marker/markers/mesh_resource_marker.cpp [hunk 1]
```diff
@@ -228,22 +235,38 @@ MeshResourceMarker::updateMaterialColor(const MarkerBase::MarkerConstSharedPtr &
...
-    technique->setAmbient(r * 0.5f, g * 0.5f, b * 0.5f);
-    technique->setDiffuse(r, g, b, a);
-    technique->setSceneBlending(blending);
-    technique->setDepthWriteEnabled(depth_write);
-    technique->setLightingEnabled(true);
+      Ogre::Technique* technique = material->getTechnique(0);
+      Ogre::Pass* pass0 = technique->getPass(0);
+      Ogre::Pass* passT = technique->getPass(technique->getNumPasses() - 1);
+      if (tinting)
+      {
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/marker/markers/mesh_resource_marker.cpp]: Actually, CI is complaining about the style here.  This should be:

```suggestion
      if (tinting) {
```

File: rviz_default_plugins/src/rviz_default_plugins/displays/marker/markers/mesh_resource_marker.cpp [hunk 2]
```diff
@@ -228,22 +235,38 @@ MeshResourceMarker::updateMaterialColor(const MarkerBase::MarkerConstSharedPtr &
...
+        // modify material's original color to use given alpha value
+        Ogre::ColourValue color = pass0->getDiffuse();
+        color.a = a;
+        pass0->setDiffuse(color);
+        // tint by re-rendering with marker color
+        passT->setAmbient(r * 0.5f, g * 0.5f, b * 0.5f);
+        passT->setDiffuse(r, g, b, std::min(a, 0.5f));
+      }
+      else
+      {
```
Comment (clalancette) [inline rviz_default_plugins/src/rviz_default_plugins/displays/marker/markers/mesh_resource_marker.cpp]:
```suggestion
      } else {
```

**Review by clalancette — CHANGES_REQUESTED** [review]
Style fixes needed.  Once those are done, we can run CI on it.

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17869/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12437/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18542/)

**Comment by xbroquere** [comment]
Hi  @clalancette ,
Is it possible to backport this change on humble ? Do I have to do something ?

Thanks

**Comment by clalancette** [comment]
> Is it possible to backport this change on humble ?

It looks like it preserves ABI and API, so yes, we can backport it.  I'll ask the bot to do it.

**Comment by clalancette** [comment]
@Mergifyio backport humble

**Pull request merged**